File: IT/Eccezioni/Esercizi/C++/esercizio28.cpp
/*
file esercizio28.cpp
compilare con: g++ -std=c++11 esercizio28.cpp
*/

#include <iostream>
#include <stdexcept>

using std::cin;
using std::cout;
using std::endl;
using std::cerr;
using std::string;
using std::to_string;
using std::logic_error;

class ATMAuthException : public logic_error {
	using logic_error::logic_error;
};

class ATM {
	private:
		bool login = false;
		bool run = true;
		double balance = 1500;
		void enterPin();
	public:
		ATM();
};

void ATM::enterPin() {
	if (!this->login) {
		cout << "Inserisci il P.I.N.: ";
		string pin;
		getline(cin, pin);

		if (pin == "12345") {
			this->login = true;
		}
		else {
			throw ATMAuthException("PIN non valido!");
		}
	}
}

ATM::ATM() {
	cout << "+++++++ Sportello ATM +++++++" << endl;

	while (this->run) {
		cout << endl << "Operazioni disponibili:" << endl << endl;
		cout << "1. Saldo" << endl;
		cout << "2. Prelievo" << endl;
		cout << "3. Deposito" << endl;
		cout << "4. Esci" << endl << endl;

		cout << "Scegli l'operazione (1, 2, 3, 4): ";

		string oper;
		getline(cin, oper);

		try {
			if (oper == "1") {
				this->enterPin();
				cout << endl << "---------------------" << endl;
				cout << "Saldo corrente: " << this->balance << " euro." << endl;
				cout << "---------------------" << endl;
			}
			else if (oper == "2") {
				this->enterPin();
				cout << "Importo da prelevare: ";

				string s;
				getline(cin, s);
				double cash = stod(s);

				this->balance -= cash;
				cout << endl << "---------------------" << endl;
				cout << "Hai prelevato " << cash << " euro. Saldo corrente: " << this->balance << endl;
				cout << "---------------------" << endl;
			}
			else if (oper == "3") {
				this->enterPin();
				cout << "Importo da versare: ";
				string s;
				getline(cin, s);
				double cash = stod(s);

				this->balance += cash;
				cout << endl << "---------------------" << endl;
				cout << "Hai versato " << cash << " euro. Saldo corrente: " << this->balance << endl;
				cout << "---------------------" << endl;
			}
			else if (oper == "4") {
				cout << "Grazie di avere utilizzato i nostri servizi ATM." << endl;
				this->run = false;
			}
			else {
				cout << "Operazione sconosciuta!" << endl;
			}
		}
		catch (logic_error &e) {
			cerr << endl << "xxxxxxxxxxxxxxxxxxxxx" << endl;
			cerr << e.what() << endl;
			cerr << "xxxxxxxxxxxxxxxxxxxxx" << endl;
		}
	}
}

int main() {
	ATM atm;
}

File: IT/1/Eccezioni/Esercizi/Java/Esercizio28.java
/* file Esercizio28.java */

import java.io.Console;

class ATMAuthException extends IllegalArgumentException {
	public ATMAuthException(String message) {
		super(message);
	}
}

class ATM {
	private boolean login = false;
	private boolean run = true;
	private double balance = 1500;
	private Console console = System.console();

	private void enterPin() throws ATMAuthException {
		if (!this.login) {
			String pin = console.readLine("Inserisci il P.I.N.: ");

			if (pin.equals("12345")) {
				this.login = true;
			}
			else {
				throw new ATMAuthException("PIN non valido!");
			}
		}
	}

	public ATM() {
		System.out.println("+++++++ Sportello ATM +++++++");

		while (this.run) {
			System.out.println("\nOperazioni disponibili:\n");
			System.out.println("1. Saldo");
			System.out.println("2. Prelievo");
			System.out.println("3. Deposito");
			System.out.println("4. Esci\n");

			String oper = console.readLine("Scegli l'operazione (1, 2, 3, 4): ");

			try {
				if (oper.equals("1")) {
					this.enterPin();
					System.out.println("\n---------------------");
					System.out.println("Saldo corrente: " + this.balance + " euro.");
					System.out.println("---------------------");
				}
				else if (oper.equals("2")) {
					this.enterPin();
					String s = console.readLine("Importo da prelevare: ");
					double cash = Double.parseDouble(s);
					this.balance -= cash;

					System.out.println("\n---------------------");
					System.out.println("Hai prelevato " + cash + " euro. Saldo corrente: " + this.balance);
					System.out.println("---------------------");
				}
				else if (oper.equals("3")) {
					this.enterPin();
					String s = console.readLine("Importo da versare: ");
					double cash = Double.parseDouble(s);
					this.balance += cash;

					System.out.println("\n---------------------");
					System.out.println("Hai versato " + cash + " euro. Saldo corrente: " + this.balance);
					System.out.println("---------------------");
				}
				else if (oper.equals("4")) {
					System.out.println("Grazie di avere utilizzato i nostri servizi ATM.");
					this.run = false;
				}
				else {
					System.out.println("Operazione sconosciuta!");
				}
			}
			catch (IllegalArgumentException e) {
				System.out.println("\nxxxxxxxxxxxxxxxxxxxxx");
				System.out.println(e.getMessage());
				System.out.println("xxxxxxxxxxxxxxxxxxxxx");
			}
		}
	}
}

public class Esercizio28 {
	public static void main(String [] args) {
		ATM atm = new ATM();
	}
}

